src/serialize.rs

```diff
@@ -1,67 +1,76 @@
 use std::io::Read;
-use std::io::Write;
+use std::io::{Error, ErrorKind};
 
 use crate::allocator::Allocator;
 use crate::node::Node;
 
 const MAX_SINGLE_BYTE: u8 = 0x7f;
 const CONS_BOX_MARKER: u8 = 0xff;
 
-fn encode_size(f: &mut dyn Write, size: usize) -> std::io::Result<()> {
+fn bad_encoding() -> std::io::Error {
+    Error::new(ErrorKind::InvalidInput, "bad encoding")
+}
+
+fn encode_size(out: &mut Vec<u8>, size: usize) {
     if size < 0x40 {
-        f.write_all(&[(0x80 | size) as u8])?;
+        out.push((0x80 | size) as u8);
     } else if size < 0x2000 {
-        f.write_all(&[(0xc0 | (size >> 8)) as u8, ((size) & 0xff) as u8])?;
+        out.extend(&[(0xc0 | (size >> 8)) as u8, ((size) & 0xff) as u8]);
     } else if size < 0x10_0000 {
-        f.write_all(&[
+        out.extend(&[
             (0xe0 | (size >> 16)) as u8,
             ((size >> 8) & 0xff) as u8,
             ((size) & 0xff) as u8,
-        ])?;
+        ]);
     } else if size < 0x800_0000 {
-        f.write_all(&[
+        out.extend(&[
             (0xf0 | (size >> 24)) as u8,
             ((size >> 16) & 0xff) as u8,
             ((size >> 8) & 0xff) as u8,
             ((size) & 0xff) as u8,
-        ])?;
+        ]);
     } else if size < 0x4_0000_0000 {
-        f.write_all(&[
+        out.extend(&[
             (0xf8 | (size >> 32)) as u8,
             ((size >> 24) & 0xff) as u8,
             ((size >> 16) & 0xff) as u8,
             ((size >> 8) & 0xff) as u8,
             ((size) & 0xff) as u8,
-        ])?;
+        ]);
+    } else {
+        panic!("atom size exceeded maximum {}", size);
     }
-    Ok(())
 }
 
-pub fn node_to_stream<T: Allocator>(node: &Node<T>, f: &mut dyn Write) -> std::io::Result<()> {
+pub fn node_to_bytes_helper<T: Allocator>(node: &Node<T>, out: &mut Vec<u8>) {
     if let Some(atom) = node.atom() {
         let size = atom.len();
         if size == 0 {
-            f.write_all(&[0x80_u8])?;
+            out.push(0x80_u8);
         } else {
             let atom0 = atom[0];
             if size == 1 && (atom0 <= MAX_SINGLE_BYTE) {
-                f.write_all(&[atom0])?;
+                out.push(atom0);
             } else {
-                encode_size(f, size)?;
-                f.write_all(&atom)?;
+                encode_size(out, size);
+                out.extend(atom);
             }
         }
     }
     if let Some((left, right)) = node.pair() {
-        f.write_all(&[CONS_BOX_MARKER as u8])?;
-        node_to_stream(&left, f)?;
-        node_to_stream(&right, f)?;
+        out.push(CONS_BOX_MARKER as u8);
+        node_to_bytes_helper(&left, out);
+        node_to_bytes_helper(&right, out);
     }
-    Ok(())
 }
 
 fn decode_size(f: &mut dyn Read, initial_b: u8) -> std::io::Result<usize> {
+    // this function decodes the length prefix for an atom. Atoms whose value
+    // fit in 7 bits don't have a length-prefix, so those should never be passed
+    // to this function.
+    assert!((initial_b & 0x80) != 0);
```

Comment:
Maybe you can return `Err` here too?

Reply:
My assumption is that asserts are compiler out in release mode, which I think is pretty reasonable for this check. If we keep it in release builds, I think it makes more sense to make it a panic. It indicates a programmer error, I don't think it should be treated as a recoverable error.

Reply:
my assumption was wrong. `assert!()` are preserved in release builds, and `debug_assert!()` is not. I think this is fine though, it should be a panic regardless